core/src/test/java/com/linecorp/armeria/client/HttpClientFactoryTest.java

```diff
@@ -103,4 +122,65 @@ protected HttpResponse doGet(ServiceRequestContext ctx,
             });
         }
     }
+
+    @Test
+    void execute_dnsTimeout_clientRequestContext_isTimedOut() {
+        try (TestDnsServer dnsServer = new TestDnsServer(ImmutableMap.of(), new AlwaysTimeoutHandler())) {
+            try (RefreshingAddressResolverGroup group = dnsTimeoutBuilder(dnsServer)
+                    .build(CommonPools.workerGroup().next())) {
+                final ClientFactory clientFactory = ClientFactory
+                        .builder()
+                        .addressResolverGroupFactory(eventExecutors -> group)
+                        .build();
+                final Endpoint endpoint = Endpoint
+                        .of("test")
+                        .withIpAddr(null); // to invoke dns resolve address
+                final WebClient client = WebClient
+                        .builder(endpoint.toUri(SessionProtocol.H1C))
+                        .factory(clientFactory)
+                        .build();
+                final HttpRequest req = HttpRequest.of(HttpMethod.GET, "/");
+                final ClientRequestContext reqCtx = ClientRequestContext
+                        .builder(req)
+                        .endpoint(endpoint)
+                        .build();
+
+                assertThat(reqCtx.isTimedOut()).isFalse();
+                assertThatThrownBy(() -> client.unwrap().execute(reqCtx, req).aggregate().join())
```

Comment:
```suggestion
                assertThatThrownBy(() -> client.get("/").aggregate().join())
```
Hmm when I use `WebClient.get("/")` directly, `reqCtx.whenResponseCancelled().join()` not done infinitely 🤔

https://github.com/line/armeria/blob/863e27c5268631360c14e50e13cd3e38d84c9db5/core/src/main/java/com/linecorp/armeria/client/ClientRequestContextBuilder.java#L144

I guess `ClientRequestContextBuilder().build()` init `cancellationScheduler` immediately, but WebClient didn't.

https://github.com/line/armeria/blob/863e27c5268631360c14e50e13cd3e38d84c9db5/core/src/main/java/com/linecorp/armeria/internal/client/DefaultClientRequestContext.java#L228-L229

https://github.com/line/armeria/blob/863e27c5268631360c14e50e13cd3e38d84c9db5/core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java#L311

On WebClient, it uses `DefaultClientRequestContext` and it doesn't `init()` the ClientRequestContext's cancellationScheduler immediately.
Just add task (e.g. `ctx.cancel(cause)`) to pending task and execute it on `init()` lazily later like below.


https://github.com/line/armeria/blob/863e27c5268631360c14e50e13cd3e38d84c9db5/core/src/main/java/com/linecorp/armeria/internal/common/CancellationScheduler.java#L129-L132

But to get `reqCtx.whenCancelled()` completion, we need to init() cancellationScheduler.

So I used `client.unwarp().execuite(reqCtx)` to use ClientRequestContext.builder().build() to init cancellationScheduler~!

I'm not sure it's correct, so feel free to share your opinion! I'll update PR 🙇

Reply:
I believe you can just use `Clients.newContextCaptor()` if you would like to get ahold of the `RequestContext` being used